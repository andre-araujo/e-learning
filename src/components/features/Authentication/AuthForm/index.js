import { Component } from 'react';

import { AlertBar, Button, TextInput } from '../../../elements';

class AuthForm extends Component {
  state = {
    name: '',
    email: '',
    password: '',
    status: undefined,
  };

  render() {
    const {
      onSubmit,
      login,
    } = this.props;

    const {
      name,
      email,
      password,
      status,
    } = this.state;

    // TODO: move to constants
    const error = (status >= 400) && 'Email ou senha inválidos';

    return (
      <form
        onSubmit={(e) => {
          e.preventDefault();
          onSubmit({ ...this.state })
            .then((resp) => {
              this.setState({ status: resp.payload.status });
              return resp;
            });
        }}
      >
        <AlertBar>
          {error}
        </AlertBar>

        {
          !login && (
            <TextInput
              label="Usuário"
              name="name"
              value={name}
              onChange={e => this.setState({
                name: e.target.value,
              })}
            />
          )
        }

        <TextInput
          label={login ? 'E-mail ou usuário' : 'E-mail'}
          name="email"
          type={login ? 'text' : 'email'}
          value={email}
          onChange={e => this.setState({
            email: e.target.value,
          })}
        />
        <TextInput
          label="Senha"
          name="password"
          value={password}
          type="password"
          onChange={e => this.setState({
            password: e.target.value,
          })}
        />
        <Button secondary full>
          Entrar
        </Button>
      </form>
    );
  }
}

export default AuthForm;
