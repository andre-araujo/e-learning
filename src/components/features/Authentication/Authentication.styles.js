import styled from 'styled-components';

export const AuthButtons = styled.div`
  max-width: 300px;
  margin: 0 auto;

  button {
    margin-bottom: 10px;
  }
`;
