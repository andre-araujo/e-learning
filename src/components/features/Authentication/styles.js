import styled from 'styled-components';

export const AuthButtons = styled.div`
    display: flex;
`;
