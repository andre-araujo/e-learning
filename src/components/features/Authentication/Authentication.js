import { connect } from 'react-redux';
import Router from 'next/router';

import AuthForm from './components/AuthForm';

import {
  Button,
  ModalProvider,
} from '../..';

import {
  AuthButtons,
} from './Authentication.styles';

function Authentication({
  requestGetToken,
  requestSingUp,
}) {
  const goToDashboard = (resp) => {
    if (resp.payload.status === 200 && resp.payload.token) {
      const isAdmin = JSON.parse(window.atob(resp.payload.token.split('.')[1])).admin;
      Router.push(isAdmin ? '/instructor-panel' : '/courses');
    }
    return resp;
  };

  return (
    <AuthButtons>
      <ModalProvider.Toggle
        title="Login"
        render={(
          <AuthForm
            login
            onSubmit={data => requestGetToken(data).then(goToDashboard)}
          />
        )}
      >
        <Button gap={15} minWidth="100px">
          Entrar
        </Button>
      </ModalProvider.Toggle>
      <ModalProvider.Toggle
        title="Cadastrar"
        render={(
          <AuthForm
            onSubmit={data => requestSingUp(data).then(goToDashboard)}
          />)}
      >
        <Button minWidth="100px">
          cadastrar
        </Button>
      </ModalProvider.Toggle>
    </AuthButtons>
  );
}

const mapStateToProps = ({ authentication }) => ({ authentication });

export default connect(mapStateToProps)(Authentication);
