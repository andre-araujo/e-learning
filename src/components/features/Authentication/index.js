import { connect } from 'react-redux';
import Router from 'next/router';

import Authentication from './Authentication';
import { actions } from '../../../lib';

const mapStateToProps = ({ authentication }) => ({ authentication });

const goToDashboard = (resp) => {
  if (resp.payload.status === 200) {
    Router.push('/my-courses');
  }
  return resp;
};

const mapDispatchToProps = dispatch => ({
  requestGetUser: () => dispatch(actions.requestGetUser()),
  requestGetToken: ({
    email,
    password,
  }) => dispatch(actions.requestGetToken({
    email,
    password,
  })).then(goToDashboard),
  requestSingUp: ({
    name,
    email,
    password,
  }) => dispatch(actions.requestSingUp({
    name,
    email,
    password,
  })).then(goToDashboard),
});

export default connect(mapStateToProps, mapDispatchToProps)(Authentication);
