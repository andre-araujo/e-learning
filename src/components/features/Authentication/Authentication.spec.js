import React from 'react';
import { mount } from 'enzyme';

import Authentication from './Authentication';

describe('Authentication component', () => {
  it('should match snapshot', () => {
    expect(mount(<Authentication />)).toMatchSnapshot();
  });
});
