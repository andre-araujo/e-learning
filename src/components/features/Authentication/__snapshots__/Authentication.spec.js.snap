// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication component should match snapshot 1`] = `
<styled.div>
  <value
    render={
      <AuthForm
        login={true}
        onSubmit={[Function]}
      />
    }
    title="Login"
  >
    <Button
      disabled={false}
      full={false}
      gap={15}
      minWidth="100px"
      secondary={false}
    >
      Entrar
    </Button>
  </value>
  <value
    render={
      <AuthForm
        onSubmit={[Function]}
      />
    }
    title="Cadastrar"
  >
    <Button
      disabled={false}
      full={false}
      gap={0}
      minWidth="100px"
      secondary={false}
    >
      cadastrar
    </Button>
  </value>
</styled.div>
`;
